tools/src/ops_tools/_generate_yaml.py

```diff
@@ -0,0 +1,434 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Generate charmcraft.yaml config and actions from Python classes."""
+
+from __future__ import annotations
+
+import dataclasses
+import enum
+import logging
+import re
+from typing import (
+    Any,
+    Final,
+    Generator,
+    Mapping,
+    TypedDict,
+    get_args,
+    get_origin,
+    get_type_hints,
+)
+
+from typing_extensions import NotRequired
+
+import ops
+
+from . import _attrdocs
+
+logger = logging.getLogger(__name__)
+
+
+class OptionDict(TypedDict):
+    type: str
+    """The Juju option type."""
+    description: NotRequired[str]
+    default: NotRequired[bool | int | float | str]
+
+
+class ActionDict(TypedDict, total=False):
+    description: str
+    params: dict[str, Any]
+    """A dictionary of parameters for the action."""
+    required: list[str]
+    """A list of required parameters for the action."""
+    additionalProperties: bool
+    """Whether additional properties are allowed in the action parameters."""
+
+
+JUJU_TYPES: Final[Mapping[type, str]] = {
+    bool: 'boolean',
+    int: 'int',
+    float: 'float',
+    str: 'string',
+    ops.Secret: 'secret',
+}
+
+# We currently only handle the basic types that we expect to see in real charms.
+# lists and tuples (arrays) are handled without using this mapping.
+JSON_TYPES: Final[Mapping[type, str]] = {
+    bool: 'boolean',
+    int: 'integer',
+    float: 'number',
+    str: 'string',
+}
+
+
+def attr_to_default(cls: type[object], name: str) -> Any:
+    """Get the default value for the attribute."""
+    if not dataclasses.is_dataclass(cls):
+        return getattr(cls, name, None)
+    default = None
+    for field in dataclasses.fields(cls):
+        if field.name != name:
+            continue
+        break
+    else:
+        return default
+
+    # This might be a Pydantic dataclass using a Pydantic.Field object.
+    field_default = (  # type: ignore
+        field.default.default  # type: ignore
+        if hasattr(field.default, 'default')
+        else field.default
+    )
+    field_default_factory = (  # type: ignore
+        field.default.default_factory  # type: ignore
+        if hasattr(field.default, 'default_factory')
+        else field.default_factory
+    )
+    # A hack to avoid importing Pydantic here.
+    if (
+        'PydanticUndefinedType' not in str(type(field_default))  # type: ignore
+        and field_default is not dataclasses.MISSING
+    ):
+        return field_default  # type: ignore
+    if field_default_factory is not dataclasses.MISSING:
+        return field_default_factory()  # type: ignore
+    return default
+
+
+def _attr_to_yaml_type(cls: type[object], name: str, yaml_types: dict[type, str]) -> str:
+    try:
+        raw_hint = get_type_hints(cls)[name]
+    except KeyError:
+        pass
+    else:
```

Comment:
```suggestion
        return 'string'
```
And dedent the rest of the function accordingly. Though here rather than having multiple occurences of the possible-to-type string value `'string'`, I would prefer using a constant, e.g. `default: Final[str] = yaml_type[str]`.